man/lightdm-webkit2-greeter.1
.TH "lightdm-webkit2-greeter" "1" "2016.01.10"
.nh
.ad l
.SH "NAME"
lightdm-webkit2-greeter
.SH "SYNOPSIS"
.PP
LightDM greeter that uses webkit2 for theming via HTML/JavaScript\&.
.PP
.SH "DESCRIPTION"
.PP
lightdm-webkit2-greeter is a LightDM greeter that uses webkit2 for theming\&.  Themes can be written
using a combination of HTML and Javascript\&.
.PP
.SH "THEME JAVASCRIPT API"
Please note that all properties and functions which are marked as "deprecated" are
only available for backwards compatibility and will be removed in a future version of
lightdm-webkit2-greeter\&.  Theme authors should not use any deprecated properties or
functions in new works and should update any existing works which make use of
deprecated properties and/or functions to ensure continued proper functionality\&.
.PP
The following functions \fBmust\fR be provided by the greeter theme and callable on the global "window" object\&.
.PP
\fBshow_prompt(text, type)\fR
.RS 4
This will be called when LightDM needs to prompt the user for some reason, such
as asking for a password\&.  The "text" parameter will be the text of the
prompt, and the "type" parameter will either be "text" for a visible prompt, or
"password" for a prompt that the input should be hidden\&.
.RE
.PP
\fBshow_message(text, type)\fR
.RS 4
This will be called when LightDM needs to display some info for the user\&.
The "text" parameter will be the text of the
message, and the "type" parameter will either be "info" for an information
message, or "error" for an error message that LightDM has encountered\&.
.RE
.PP
\fBauthentication_complete()\fR
.RS 4
This function is called by LightDM when authentication has completed\&.
.RE
.PP
\fBautologin_timer_expired()\fR
.RS 4
This function is called by LightDM when an autologin user's login timer has
expired\&.  The greeter should reset the authentication process\&.
.RE
.PP
The following functions are available for the greeter to call to execute
actions within LightDM\&.
.PP
\fBlightdm\&.authenticate(username)\fR
.RS 4
Specifies the username of the user we'd like to start authenticating as\&.  Note that
if you call lightdm.authenticate with no argument, LightDM (via PAM) will issue
a show_prompt() call to ask for the username\&. The older function
lightdm\&.start_authentication() has been deprecated\&.
.RE
.PP
\fBlightdm\&.authenticate_as_guest()\fR
.RS 4
Authenticates as the guest user\&.
.RE
.PP
\fBlightdm\&.cancel_authentication()\fR
.RS 4
Cancels the authentication of any user currently in the process of
authenticating\&.
.RE
.PP
\fBlightdm\&.cancel_autologin()\fR
.RS 4
Cancels the authentication of the autologin user\&.  The older function
lightdm\&.cancel_timed_login() has been deprecated.
.RE
.PP
\fBlightdm\&.get_hint(hint_name)\fR
.RS 4
Returns the value of a named hint provided by LightDM\&.
.RE
.PP
\fBlightdm\&.hibernate()\fR
.RS 4
Hibernates the system, if the greeter has the authority to do so\&.
.RE
.PP
\fBlightdm\&.start_session_sync(session)\fR
.RS 4
Once LightDM has successfully authenticated the user, start the user's session
by calling this function\&.  "session" is the authenticated user's session\&.
If no session is passed, start the authenticated user with the system default
session. The older function lightdm\&.login(user, session) has been
deprecated\&.
.RE
.PP
\fBlightdm\&.respond(text)\fR
.RS 4
When LightDM has prompted for input, provide the response to LightDM\&.
The deprecated function was "provide_secret"\&.  This is still available for
backwards compatibility, but authors of greeters should move to using
lightdm.respond()\&.
.RE
.PP
\fBlightdm\&.restart()\fR
.RS 4
Restarts the system, if the greeter has the authority to do so\&.
.RE
.PP
\fBlightdm\&.set_language(lang)\fR
.RS 4
Will set the language for the current LightDM session\&.
.RE
.PP
\fBlightdm\&.shutdown()\fR
.RS 4
Shuts down the system, if the greeter has the authority to do so\&.
.RE
.PP
Variables available within the greeter are:
.PP
\fBlightdm\&.authentication_user\fR
.RS 4
String\&. The username of the authentication user being authenticated or null if no
authentication is in progress\&.
.RE
.PP
\fBlightdm\&.autologin_guest\fR
.RS 4
Boolean\&. Indicates the guest user should be used for autologin\&.
.RE
.PP
\fBlightdm\&.autologin_timeout\fR
.RS 4
Number\&. The number of seconds to wait before automatically logging in\&. The
older variable lightdm\&.timed_user_delay has been deprecated\&.
.RE
.PP
\fBlightdm\&.autologin_user\fR
.RS 4
String\&. The name of the user account that should be logged into
automatically after timed login delay has passed\&. The older variable
lightdm\&.timed_login_user has been deprecated\&.
.RE
.PP
\fBlightdm\&.can_hibernate\fR
.RS 4
Boolean\&. Whether or not the system can be made to hibernate by the greeter\&.
.RE
.PP
\fBlightdm\&.can_restart\fR
.RS 4
Boolean\&. Whether or not the system can be restarted by the greeter\&.
.RE
.PP
\fBlightdm\&.can_shutdown\fR
.RS 4
Boolean\&. Whether or not the system can be shutdown by the greeter\&.
.RE
.PP
\fBlightdm\&.can_suspend\fR
.RS 4
Boolean\&. Whether or not the system can be suspended by the greeter\&.
.RE
.PP
\fBlightdm\&.default_session\fR
.RS 4
String\&. The name of the default session (as configured in lightdm.conf)\&.
.RE
.PP
\fBlightdm\&.has_guest_account\fR
.RS 4
Boolean\&. A guest account is available for login\&.
.RE
.PP
\fBlightdm\&.hide_users\fR
.RS 4
Boolean\&. The whole list of users should not be displayed\&.
.RE
.PP
\fBlightdm\&.hostname\fR
.RS 4
String\&. The hostname of the system\&.
.RE
.PP
\fBlightdm\&.is_authenticated\fR
.RS 4
Boolean\&. Indicates if the user has successfully authenticated\&.
.RE
.PP
\fBlightdm\&.in_authentication\fR
.RS 4
Boolean\&. Indicates if lightdm is currently in the authentication phase\&.
.RE
.PP
\fBlightdm\&.language\fR
.RS 4
String\&. The currently selected language\&. The older variable name
lightdm\&.default_language is deprecated\&.
.RE
.PP
\fBlightdm\&.layout\fR
.RS 4
String\&. The name of the currently active keyboard layout\&. To change the
layout, assign a valid layout name to this variable\&. The older variable name
lightdm\&.default_layout is deprecated\&.
.RE
.PP
\fBlightdm\&.layouts\fR
.RS 4
Array\&. The keyboard layouts that are available on the system\&. Returns an Array
of LightDMLayout objects\&.
.RE
.PP
\fBlightdm\&.num_users\fR
.RS 4
Number\&. The number of users able to log in\&.
.RE
.PP
\fBlightdm\&.select_guest\fR
.RS 4
Boolean\&. The guest user should be selected by default for login\&.
.RE
.PP
\fBlightdm\&.select_user\fR
.RS 4
String\&. The username that should be selected by default for login\&.
.RE
.PP
\fBlightdm\&.sessions\fR
.RS 4
Array\&. The sessions that are available on the system\&. Returns an Array of
LightDMSession objects\&.
.RE
.PP
\fBlightdm\&.users\fR
.RS 4
Array\&. The users that are able to log in\&. Returns an Array of LightDMUser
objects\&.
.RE
.PP
The following calls can be made to read configuration keys out of the
lightdm-webkit2-greeter configuration file\&.
.PP
config\&.get_str(section, key)\fR
.RS 4
Returns the string value associated with key under the "section" in the configuration file\&.
.RE
.PP
config\&.get_num(section, key)\fR
.RS 4
Returns the numeric value associated with key under the "section" in the configuration file\&.
.RE
.PP
config\&.get_bool(section, key)\fR
.RS 4
Returns the boolean value associated with key under the "section" in the configuration file\&.
.RE
.PP
The greeterutil object has some utility functions associated with it which
are intended to make a theme author's work easier\&.
.PP
greeterutil\&.dirlist(path)\fR
.RS 4
Returns an array of strings of filenames present at "path", or Null if the
path does not exist\&.
.RE
.PP
greeterutil\&.txt2html(txt)\fR
.RS 4
Returns a simple HTML conversion of the passed text\&.  Newlines are replaced
with <br>, and the characters \&&, <, >, and " are replaced with their HTML
equivalents\&.
.RE
.PP
Please see the LightDM API documentation for the complete list of calls
available\&.  The lightdm-webkit2-greeter implements all of the LightDM API\&.
.PP
.SH "CONFIGURATION"
.PP
\fB/etc/lightdm/lightdm-webkit2-greeter\&.conf\fR
.RS 4
Keyfile that contains one key: webkit-theme\&.  This should point to which
theme the greeter should use\&.
.RE
.SH "FILES"
.PP
\fB/usr/share/lightdm-webkit/themes\fR
.RS 4
Directory where themes should be stored\&.
.RE
.SH "EXAMPLES"
.PP
Please see the "antergos" and "simple" themes that are shipped with lightdm-webkit2-greeter\&.
.SH "SEE ALSO"
.PP
http://people\&.ubuntu\&.com/~robert-ancell/lightdm/reference/core\&.html
.SH "AUTHOR"
.PP
The legacy lightdm-webkit-greeter was written by Robert Ancell <robert\&.ancell\&@canonical\&.com\&>\&.
It was ported to webkit2 by the Antergos Developers <dev@antergos\&.com>\&. Also includes code improvements
contributed by Scott Balneaves <sbalneav@ltsp\&.org>\&.


NEWS
Overview of changes in lightdm-webkit2-greeter 2.0.0

    * Version scheme updated for both the webkit1 and webkit2 greeters as follows:
        - MAJOR.MINOR.UNIQUE_FIXES
            * MAJOR: Denotes a major release. Shared by legacy webkit1 and webkit2 greeters.
            * MINOR: Denotes small changes/fixes. Also shared by webkit 1 and webkit2 greeters.
            * UNIQUE_FIXES: Denotes fixes unique to either the legacy webkit1 or webkit2 greeters.
    * Legacy Webkit1 greeter development is officially in maintenance-only mode (bug fixes only).
    * Webkit2 greeter supersedes the Webkit1 greeter and is the focus of all new development.
    * Completed API including the guest and auto-login functions.
    * API is now completely in-line with LightDM API.
    * Several functional code improvements.
    * Added Exception handling.
    * Added man page.
    * Default theme:
        - Added translations.
        - Added confirmation dialogue for system actions.
        - Theme's script was completely rewritten in ES6.
        - Use babel to compile ES6->ES5 until webkit catches up.

Overview of changes in lightdm-webkit2-greeter 0.2.3

    * Add support for getting the lock_hint from lightdm.
    * Default webkit theme is now antergos.
    * Code formatting cleanup

Overview of changes in lightdm-webkit2-greeter 0.2.1

    * Handle tasks that a WM would normally handle.
    * Code clean-up

Overview of changes in lightdm-webkit2-greeter 0.2.0

    * Add missing lightdm.set_language function
    * Fix compile warnings

Overview of changes in lightdm-webkit2-greeter 0.1.3

    * Port from webkitgtk -> webkit2gtk
    * Use GTK+ 3.0

Overview of changes in lightdm-webkit-greeter 0.1.2

    * Fix files not being added to tarball
    * Don't use AC_CONFIG_MACRO_DIR in configure.ac

Overview of changes in lightdm-webkit-greeter 0.1.1

    * Update to work with newer LightDM

Overview of changes in lightdm-webkit-greeter 0.1.0

    * Split out into separate module from lightdm


src/lightdm-webkit2-greeter.c
/*
 * lightdm-webkit2-greeter.c
 *
 * Copyright © 2014-2016 Antergos Developers <dev@antergos.com>
 *
 * Includes Code Contributed By:
 * Copyright © 2016 Scott Balneaves <sbalneav@ltsp.org>
 *
 * Based on code from lightdm-webkit-greeter:
 * Copyright © 2010-2015 Robert Ancell <robert.ancell@canonical.com>
 *
 * This file is part of lightdm-webkit2-greeter.
 *
 * lightdm-webkit2-greeter is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * lightdm-webkit2-greeter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * The following additional terms are in effect as per Section 7 of the license:
 *
 * The preservation of all legal notices and author attributions in
 * the material or in the Appropriate Legal Notices displayed
 * by works containing it is required.
 *
 * You should have received a copy of the GNU General Public License
 * along with lightdm-webkit2-greeter; If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdlib.h>
#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#include <glib.h>
#include <glib-unix.h>
#include <gtk/gtkx.h>
#include <webkit2/webkit2.h>
#include <JavaScriptCore/JavaScript.h>
#include <glib/gi18n.h>
#include <sys/mman.h>

#include <lightdm.h>

#include <config.h>

#include "src/lightdm-webkit2-greeter-css-application.h"

static GtkWidget *web_view;
static GtkWidget *window;
static WebKitSettings *webkit_settings;
static GdkDisplay *default_display;

/* Screensaver values */
static int timeout, interval, prefer_blanking, allow_exposures;
static gint config_timeout;


static GdkFilterReturn
wm_window_filter(GdkXEvent *gxevent, GdkEvent *event, gpointer data) {
	XEvent *xevent = (XEvent *) gxevent;

	if (xevent->type == MapNotify) {
		GdkDisplay *display = gdk_x11_lookup_xdisplay(xevent->xmap.display);
		GdkWindow *win = gdk_x11_window_foreign_new_for_display(display, xevent->xmap.window);
		GdkWindowTypeHint win_type = gdk_window_get_type_hint(win);

		if (win_type != GDK_WINDOW_TYPE_HINT_COMBO
			&& win_type != GDK_WINDOW_TYPE_HINT_TOOLTIP
			&& win_type != GDK_WINDOW_TYPE_HINT_NOTIFICATION) {

			gdk_window_focus(win, GDK_CURRENT_TIME);
		}

	} else if (xevent->type == UnmapNotify) {
		Window xwin;
		int revert_to = RevertToNone;

		XGetInputFocus(xevent->xunmap.display, &xwin, &revert_to);
		if (revert_to == RevertToNone) {
			gdk_window_lower(gtk_widget_get_window(gtk_widget_get_toplevel(GTK_WIDGET(window))));
		}
	}

	return GDK_FILTER_CONTINUE;
}


static void
initialize_web_extensions_cb(WebKitWebContext *context, gpointer user_data) {

	webkit_web_context_set_web_extensions_directory(context, LIGHTDM_WEBKIT2_GREETER_EXTENSIONS_DIR);

}


static void
create_new_webkit_settings_object(void) {
	webkit_settings = webkit_settings_new_with_settings(
		"enable-developer-extras", FALSE,
		"enable-fullscreen", TRUE,
		"enable-site-specific-quirks", TRUE,
		"enable-dns-prefetching", TRUE,
		"javascript-can-open-windows-automatically", TRUE,
		"allow-file-access-from-file-urls", TRUE,
		"enable-accelerated-2d-canvas", TRUE,
		"enable-smooth-scrolling", FALSE,
		"enable-webgl", TRUE,
		"enable-write-console-messages-to-stdout", TRUE,
		NULL
	);
}


static gboolean
context_menu_cb(WebKitWebView *view,
				WebKitContextMenu *context_menu,
				GdkEvent *event,
				WebKitHitTestResult *hit_test_result,
				gpointer user_data) {

	return TRUE;
}


/**
 * Lock Hint enabled handler.
 *
 * Makes the greeter behave a bit more like a screensaver if it was launched as
 * a lock-screen by blanking the screen.
 */
static void
lock_hint_enabled_handler(void) {
	Display *display = gdk_x11_display_get_xdisplay(default_display);

	XGetScreenSaver(display, &timeout, &interval, &prefer_blanking, &allow_exposures);
	XForceScreenSaver(display, ScreenSaverActive);
	XSetScreenSaver(display, config_timeout, 0, ScreenSaverActive, DefaultExposures);
}


/**
 * Message received callback.
 *
 * Receives messages from our web extension process and calls appropriate handlers.
 *
 * @param: manager The WebKitUserContentManager instance that was created in #main.
 * @param: message The message sent from web extension process.
 * @param: user_data Data that is private to the current user.
 */
static void
message_received_cb(WebKitUserContentManager *manager,
					WebKitJavascriptResult *message,
					gpointer user_data) {

	/* TODO:
	 * Abstract this by using JSON for exchanging messages so the handler can
	 * be used for more than one task/event.
	 */
	lock_hint_enabled_handler();

}


int
main(int argc, char **argv) {
	GdkScreen *screen;
	GdkWindow *root_window;
	GdkRectangle geometry;
	GKeyFile *keyfile;
	gchar *theme;
	GdkRGBA bg_color;
	WebKitUserContentManager *manager;
	WebKitWebContext *context;
	GtkCssProvider *css_provider;

	/*
	 * Prevent memory from being swapped out, since we see unencrypted
	 * passwords.
	 */
	mlockall (MCL_CURRENT | MCL_FUTURE);

	/* Initialize i18n */
	setlocale (LC_ALL, "");
	bindtextdomain (GETTEXT_PACKAGE, LOCALE_DIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	gtk_init(&argc, &argv);
	g_unix_signal_add(SIGTERM, gtk_main_quit, NULL);

	/* Apply greeter settings from config file */
	keyfile = g_key_file_new();

	g_key_file_load_from_file(keyfile,
							  CONFIG_DIR "/lightdm-webkit2-greeter.conf",
							  G_KEY_FILE_NONE, NULL);

	theme = g_key_file_get_string(keyfile, "greeter", "webkit-theme", NULL);
	config_timeout = g_key_file_get_integer(keyfile, "greeter", "screensaver-timeout", NULL);

	/* Setup the main window */
	window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
	screen = gtk_window_get_screen(GTK_WINDOW(window));
	root_window = gdk_get_default_root_window();
	default_display = gdk_display_get_default();

	gtk_window_set_decorated(GTK_WINDOW(window), FALSE);
	gdk_screen_get_monitor_geometry(screen, gdk_screen_get_primary_monitor(screen), &geometry);
	gtk_window_set_default_size(GTK_WINDOW(window), geometry.width, geometry.height);
	gtk_window_move(GTK_WINDOW(window), geometry.x, geometry.y);

	/* There is no window manager, so we need to implement some of its functionality */
	gdk_window_set_events(root_window, gdk_window_get_events(root_window) | GDK_SUBSTRUCTURE_MASK);
	gdk_window_add_filter(root_window, wm_window_filter, NULL);

	/* Setup CSS provider. We use CSS to set the window background to black instead
	 * of default white so the screen doesnt flash during startup.
	 */
	css_provider = gtk_css_provider_new();
	gtk_css_provider_load_from_data(css_provider,
									lightdm_webkit2_greeter_css_application,
									lightdm_webkit2_greeter_css_application_length, NULL);
	gtk_style_context_add_provider_for_screen(screen,
											  GTK_STYLE_PROVIDER(css_provider),
											  GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);

	/* Register and connect handler that will set the web extensions directory
	 * so webkit can find our extension.
	 */
	context = webkit_web_context_get_default();
	g_signal_connect(context,
					 "initialize-web-extensions",
					 G_CALLBACK(initialize_web_extensions_cb), NULL);

	/* Register and connect handler of any messages we send from our web extension process. */
	manager = webkit_user_content_manager_new();
	webkit_user_content_manager_register_script_message_handler(manager, "GreeterBridge");
	g_signal_connect(manager,
					 "script-message-received::GreeterBridge",
					 G_CALLBACK(message_received_cb), NULL);

	/* Create the web_view */
	web_view = webkit_web_view_new_with_user_content_manager(manager);

	/* Set the web_view's settings. */
	create_new_webkit_settings_object();
	webkit_web_view_set_settings(WEBKIT_WEB_VIEW(web_view), webkit_settings);

	/* The default background is white which causes a flash effect when the greeter starts.
	 * We make it black instead. This is for backwards compatibility with Gtk versions that
	 * don't use the new CSS provider.
	 */
	gdk_rgba_parse(&bg_color, "#000000");
	webkit_web_view_set_background_color(WEBKIT_WEB_VIEW(web_view), gdk_rgba_copy(&bg_color));

	/* Disable the context (right-click) menu. */
	g_signal_connect(web_view, "context-menu", G_CALLBACK(context_menu_cb), NULL);

	/* There's no turning back now, let's go! */
	gtk_container_add(GTK_CONTAINER(window), web_view);
	webkit_web_view_load_uri(WEBKIT_WEB_VIEW(web_view),
							 g_strdup_printf("file://%s/%s/index.html", THEME_DIR, theme));

	gtk_widget_show_all(window);
	gdk_window_set_cursor(gtk_widget_get_window (GTK_WIDGET (window)), gdk_cursor_new_for_display(default_display, GDK_LEFT_PTR));

	gtk_main();

	return 0;
}
